libs/langgraph/src/pregel/remote.ts

```diff
@@ -0,0 +1,446 @@
+
+import {
+  BaseChannel,
+  LangGraphRunnableConfig,
+  ManagedValueSpec,
+} from "../web.js";
+import { StrRecord } from "./algo.js";
+import { PregelInputType, PregelOptions, PregelOutputType } from "./index.js";
+import { PregelNode } from "./read.js";
+import {
+  PregelParams,
+  PregelInterface,
+  PregelTaskDescription,
+  StateSnapshot,
+  StreamMode,
+} from "./types.js";
+import { Interrupt } from "../constants.js";
+
+export type RemoteGraphParams = Omit<
+  PregelParams<
+    StrRecord<string, PregelNode>,
+    StrRecord<string, BaseChannel | ManagedValueSpec>
+  >,
+  "channels" | "nodes" | "inputChannels" | "outputChannels"
+> & {
+  graphId: string;
+  client: Client;
+};
+
+export class RemoteGraph<
+    Nn extends StrRecord<string, PregelNode> = StrRecord<string, PregelNode>,
+    Cc extends StrRecord<string, BaseChannel | ManagedValueSpec> = StrRecord<
+      string,
+      BaseChannel | ManagedValueSpec
+    >,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    ConfigurableFieldType extends Record<string, any> = StrRecord<string, any>
+  >
+  extends Runnable<
+    PregelInputType,
+    PregelOutputType,
+    PregelOptions<Nn, Cc, ConfigurableFieldType>
+  >
+  implements PregelInterface<Nn, Cc, ConfigurableFieldType>
+{
+  static lc_name() {
+    return "RemoteGraph";
+  }
+
+  lc_namespace = ["langgraph", "pregel"];
+
+  lg_is_pregel = true;
+
+  config?: RunnableConfig;
+
+  graphId: string;
+
+  protected client: Client;
+
+  protected interruptBefore?: Array<keyof Nn> | All;
+
+  protected interruptAfter?: Array<keyof Nn> | All;
+
+  constructor(params: RemoteGraphParams) {
+    super(params);
+
+    this.graphId = params.graphId;
+    this.client = params.client;
+    this.config = params.config;
+    this.interruptBefore = params.interruptBefore;
+    this.interruptAfter = params.interruptAfter;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore Remove ignore when we remove support for 0.2 versions of core
+  override withConfig(config: RunnableConfig): typeof this {
+    const mergedConfig = mergeConfigs(this.config, config);
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return new (this.constructor as any)({ ...this, config: mergedConfig });
+  }
+
+  protected _sanitizeConfig(config: RunnableConfig) {
+    const reservedConfigurableKeys = new Set([
+      "callbacks",
+      "checkpoint_map",
+      "checkpoint_id",
+      "checkpoint_ns",
+    ]);
+
+    const sanitizeObj = (obj: any): any => {
+      // Remove non-JSON serializable fields from the given object
+      if (obj && typeof obj === "object") {
+        if (Array.isArray(obj)) {
+          return obj.map((v) => sanitizeObj(v));
+        } else {
+          return Object.fromEntries(
+            Object.entries(obj).map(([k, v]) => [k, sanitizeObj(v)])
+          );
+        }
+      }
+
+      try {
+        JSON.stringify(obj);
```

Comment:
_question_: so the behavior is to bail out when we encounter a circular dependency and return `null` instead? Is this correct behaviour?

Reply:
We could do the same thing as in `langsmith` if we wanted

Reply:
I will add a TODO

Reply:
Actually no I think this is correct